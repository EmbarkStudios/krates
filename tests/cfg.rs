mod util;

use krates::cfg_expr::targets;
use util::{build, cmp};

#[test]
fn ignores_non_linux() {
    let mut kb = krates::Builder::new();
    kb.include_targets(targets::ALL_BUILTINS.iter().filter_map(|ti| {
        if ti.os == Some(targets::Os::linux) {
            Some((ti.triple.clone(), Vec::new()))
        } else {
            None
        }
    }));

    let grafs = build("all-features.json", kb).unwrap();

    let targets: Vec<_> = targets::ALL_BUILTINS
        .iter()
        .filter(|ti| ti.os == Some(targets::Os::linux))
        .collect();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[test]
fn ignores_non_tier1() {
    let mut kb = krates::Builder::new();

    let targets = vec![
        targets::get_builtin_target_by_triple("i686-pc-windows-gnu").unwrap(),
        targets::get_builtin_target_by_triple("i686-pc-windows-msvc").unwrap(),
        targets::get_builtin_target_by_triple("i686-unknown-linux-gnu").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-apple-darwin").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-pc-windows-gnu").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-pc-windows-msvc").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-unknown-linux-gnu").unwrap(),
    ];

    kb.include_targets(targets.iter().map(|ti| (ti.triple.clone(), vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[test]
fn ignores_non_wasm() {
    let mut kb = krates::Builder::new();

    let targets = vec![targets::get_builtin_target_by_triple("wasm32-unknown-unknown").unwrap()];

    kb.include_targets(targets.iter().map(|ti| (ti.triple.clone(), vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[cfg(feature = "targets")]
#[test]
fn handles_non_builtin() {
    let mut kb = krates::Builder::new();

    use krates::target_lexicon::Triple;

    let xbox: Triple = "x86_64-xboxone-windows-msvc".parse().unwrap();

    kb.include_targets(std::iter::once(("x86_64-xboxone-windows-msvc", vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    !expr.eval(|pred| match pred {
                        krates::cfg_expr::Predicate::Target(tp) => tp.matches(&xbox),
                        _ => false,
                    })
                } else {
                    "x86_64-xboxone-windows-msvc" != cfg
                }
            } else {
                false
            }
        },
    );
}
