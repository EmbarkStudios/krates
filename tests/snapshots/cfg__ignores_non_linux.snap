---
source: tests/cfg.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 path+file:///krates/tests/ws/a" ]
    1 [ label = "crate b 0.1.0 path+file:///krates/tests/ws/b" ]
    2 [ label = "crate bitflags 1.2.1" ]
    3 [ label = "crate c 0.1.0 path+file:///krates/tests/ws/c" ]
    4 [ label = "crate cc 1.0.50" ]
    5 [ label = "crate cc 1.0.84 git+https://github.com/alexcrichton/cc-rs?branch=main" ]
    6 [ label = "crate cc 1.0.84 git+https://github.com/alexcrichton/cc-rs?rev=34d4ce4" ]
    7 [ label = "crate cfg-if 0.1.10" ]
    8 [ label = "crate difference 2.0.0" ]
    9 [ label = "crate lazy_static 1.4.0" ]
    10 [ label = "crate leftpad 0.2.0" ]
    11 [ label = "crate libc 0.2.66" ]
    12 [ label = "crate nix 0.16.1" ]
    13 [ label = "crate ring 0.16.9" ]
    14 [ label = "crate spin 0.5.2" ]
    15 [ label = "crate untrusted 0.7.0" ]
    16 [ label = "crate void 1.0.2" ]
    17 [ label = "feature default" ]
    18 [ label = "feature default" ]
    19 [ label = "feature default" ]
    20 [ label = "feature default" ]
    21 [ label = "feature extra_traits" ]
    22 [ label = "feature default" ]
    23 [ label = "feature default" ]
    24 [ label = "feature leftpad" ]
    25 [ label = "feature leftier-strings" ]
    26 [ label = "feature lazy_static" ]
    27 [ label = "feature std" ]
    28 [ label = "feature lazy_static" ]
    29 [ label = "feature dev_urandom_fallback" ]
    30 [ label = "feature alloc" ]
    31 [ label = "feature std" ]
    0 -> 1 [ label = "" ]
    0 -> 17 [ label = "(dev)" ]
    0 -> 17 [ label = "(build) 'cfg(target_os = \"linux\")'" ]
    1 -> 17 [ label = "" ]
    1 -> 6 [ label = "(build)" ]
    1 -> 18 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    3 -> 5 [ label = "(build)" ]
    3 -> 19 [ label = "(dev)" ]
    3 -> 9 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    3 -> 10 [ label = "" ]
    3 -> 11 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    3 -> 12 [ label = " 'x86_64-unknown-linux-gnu'" ]
    3 -> 14 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    5 -> 11 [ label = " 'cfg(unix)'" ]
    6 -> 11 [ label = " 'cfg(unix)'" ]
    12 -> 20 [ label = "" ]
    12 -> 7 [ label = "" ]
    12 -> 21 [ label = "" ]
    12 -> 22 [ label = "" ]
    12 -> 23 [ label = "" ]
    13 -> 4 [ label = "(build)" ]
    13 -> 9 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    13 -> 11 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    13 -> 14 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    13 -> 15 [ label = "" ]
    20 -> 2 [ label = "" ]
    24 -> 3 [ label = "" ]
    24 -> 10 [ label = "" ]
    25 -> 3 [ label = "" ]
    25 -> 24 [ label = "" ]
    26 -> 3 [ label = "" ]
    26 -> 9 [ label = "" ]
    17 -> 3 [ label = "" ]
    17 -> 24 [ label = "" ]
    19 -> 8 [ label = "" ]
    27 -> 11 [ label = "" ]
    21 -> 11 [ label = "" ]
    22 -> 11 [ label = "" ]
    22 -> 27 [ label = "" ]
    28 -> 13 [ label = "" ]
    28 -> 9 [ label = "" ]
    29 -> 13 [ label = "" ]
    29 -> 28 [ label = "" ]
    18 -> 13 [ label = "" ]
    18 -> 30 [ label = "" ]
    18 -> 29 [ label = "" ]
    30 -> 13 [ label = "" ]
    31 -> 16 [ label = "" ]
    23 -> 16 [ label = "" ]
    23 -> 31 [ label = "" ]
}

